presto-rcfile/src/main/java/com/facebook/presto/rcfile/RcFileReader.java

```diff
@@ -0,0 +1,584 @@
+        int uncompressedHeaderSize = Integer.reverseBytes(input.readInt());
+        int compressedHeaderSize = Integer.reverseBytes(input.readInt());
+        if (codec == null && compressedHeaderSize != uncompressedHeaderSize) {
+            close();
+            throw corrupt("Invalid RCFile %s", dataSource);
+        }
+        if (compressedHeaderSize > compressedHeaderBuffer.length()) {
+            compressedHeaderBuffer = Slices.allocate(compressedHeaderSize);
+        }
+        input.readBytes(compressedHeaderBuffer, 0, compressedHeaderSize);
+
+        // decompress block header
+        Slice header;
+        if (codec != null) {
+            if (headerBuffer.length() < uncompressedHeaderSize) {
+                headerBuffer = Slices.allocate(uncompressedHeaderSize);
+            }
+            Slice buffer = headerBuffer.slice(0, uncompressedHeaderSize);
+
+            codec.decompress(compressedHeaderBuffer, buffer);
+
+            header = buffer;
+        }
+        else {
+            header = compressedHeaderBuffer.slice(0, uncompressedHeaderSize);
+        }
+        BasicSliceInput headerInput = header.getInput();
+
+        // read number of rows in block
+        blockRowCount = Ints.checkedCast(readVInt(headerInput));
+        rowsRead += blockRowCount;
+        blockPosition = 0;
+        currentChunkSize = min(ColumnData.MAX_SIZE, blockRowCount);
+
+        // set column buffers
+        for (int columnIndex = 0; columnIndex < columnCount; columnIndex++) {
+            int compressedDataSize = Ints.checkedCast(readVInt(headerInput));
+            int uncompressedDataSize = Ints.checkedCast(readVInt(headerInput));
+            if (codec == null && compressedDataSize != uncompressedDataSize) {
+                throw corrupt("Invalid RCFile %s", dataSource);
+            }
+
+            int lengthsSize = Ints.checkedCast(readVInt(headerInput));
+
+            Slice lengthsBuffer = headerInput.readSlice(lengthsSize);
+
+            Slice dataBuffer;
+            if (readColumns.contains(columnIndex)) {
+                dataBuffer = input.readSlice(compressedDataSize);
+            }
+            else {
+                input.skip(compressedDataSize);
+                dataBuffer = Slices.EMPTY_SLICE;
+            }
+
+            columns[columnIndex].setBuffers(lengthsBuffer, dataBuffer, uncompressedDataSize);
+        }
+
+        return currentChunkSize;
+    }
+
+    public ColumnData readColumnData(int columnIndex)
+            throws IOException
+    {
+        if (currentChunkSize <= 0) {
+            throw new IllegalStateException("No more data");
+        }
+        return columns[columnIndex].readColumnData(blockPosition, currentChunkSize);
+    }
+
+    private void seekToFirstBlockIn(long offset, long length)
+            throws IOException
+    {
+        if (length < SYNC_SEQUENCE_LENGTH) {
+            closeQuietly();
+            return;
+        }
+
+        Slice sync = getSync();
```

Comment:
The required `Int -1` before `SYNC_SEQUENCE` should/could be part of the needle you use to find the next row group